Starship fixes (#148)
XanaduAI/strawberryfields: doc/tutorials/starship.rst

.. _starship:

Running Jobs with StarshipEngine
################################

.. sectionauthor:: Zeid Zabaneh <zeid@xanadu.ai>

> In this section, we provide a tutorial of the **StarshipEngine**, an engine used to connect to the Strawberry Fields platform and execute jobs remotely (e.g. on a quantum chip).

Reviewer: ```suggestion
In this section, we provide a tutorial of the **StarshipEngine**, an engine used to connect to the Xanadu cloud platform and execute jobs remotely (e.g., on a quantum chip).
```

Configuring StarshipEngine
==========================

> Before using StarshipEngine, you need to configure the hostname and authentication token that will provide you access to the API. The easiest way is to create a configuration file named ``config.toml`` in your working directory. A typical file looks like this:

Reviewer: This works now in local directory? I may have missed it, but didn't see where the code was changed to look for this locally

Author: By locally do you mean current working directory? It has always looked there first, but I forgot about that feature last week.

Reviewer: Yep, that's what I meant. Could have saved a lot of hassle with different operating systems if I knew you could just put it there!

Author: Yes for sure. It is in the documentation as the first point, but it completely did not occur to me to use it then.

.. code-block:: console

   [api]
   hostname = "platform.example.com"
   authentication_token = "ElUFm3O6m6q1DXPmpi5g4hWEhYHXFxBc"

You can generate this file interactively by using the ``starship`` command as follows, answering the questions in the prompts.

.. code-block:: console

>    (strawberryfields)$ starship --reconfigure

Reviewer: What is `(strawberryfields)`? an environment?

Author: Yes - I will change this to be more generic though, because it doesn't make sense this way.

Reviewer 2: It might be worth removing the `$` as well. We used to use it everywhere to denote a shell environment (and I think it is still used on the installation page), but I've had some users copy and paste the dollar sign thinking it was part of the command.

   Please enter the hostname of the server to connect to: [platform.example.com]
   Please enter the authentication token to use when connecting: [] ElUFm3O6m6q1DXPmpi5g4hWEhYHXFxBc
   Would you like to save these settings to a local cofiguration file in the current directory? [Y/n] y
   Writing configuration file to current working directory...

To test connectivity, you can use the following command:

.. code-block:: console

>    (strawberryfields)$ starship --hello

Reviewer: Nice, agreed this would be useful to have. Would make tests easier too

Reviewer 2: I like the `--hello` flag :laughing:

   You have successfully authenticated to the platform!

Executing your first program
============================

> The easiest way to execute a program using StarshipEngine is to create an ``xbb`` file and place it in your current working directory. Check the :ref:`blackbird tutorial <tutorial>` for how to create this file.

Reviewer: ```suggestion
The easiest way to execute a program using StarshipEngine is to create a blackbird script (an ``xbb`` file) and place it in your current working directory. Check the :ref:`blackbird tutorial <tutorial>` for how to create this file.
```

For this example, we will use the following file and save it to ``test.xbb`` in our current working directory:

> .. code-block::

Reviewer 2: This won't display in the docs, code-block requires a language specified. You can just use Python, as the Blackbird syntax started as a Python DSL anyway (so it works well)

```suggestion
.. code-block:: python
```

Author: Compiling it locally does work, the syntax highlighting is a little different though.

Reviewer 2: That's strange, we had the opposite experience with the Sphinx environment on read the docs the other week (see https://github.com/XanaduAI/pennylane/issues/249)

   name template_2x2_chip0
   version 1.0
   target chip0 (shots = 50)

   float sqz0 = 1.0
   float sqz1 = 1.0
   float phi0 = 0.574
   float phi1 = 1.33

   # for n spatial degrees, first n signal modes, then n idler modes, phase zero
   S2gate(sqz0, 0.0) | [0, 2]
   S2gate(sqz1, 0.0) | [1, 3]

   # standard 2x2 interferometer for the signal modes (the lower ones in frequency)
   Rgate(phi0) | [0]
   BSgate(pi/4, pi/2) | [0, 1]
   Rgate(phi1) | [0]
   BSgate(pi/4, pi/2) | [0, 1]

   #duplicate the interferometer for the idler modes (the higher ones in frequency)
   Rgate(phi0) | [2]
   BSgate(pi/4, pi/2) | [2, 3]
   Rgate(phi1) | [2]
   BSgate(pi/4, pi/2) | [2, 3]

   # Measurement in Fock basis
   MeasureFock() | [0, 1, 2, 3]

After you have created your ``xbb`` file, you can execute it using the command line, or using a python shell.

Executing your xbb file using Python
====================================

> To execute this file from using Python, you can use a code block like this:

Reviewer: ```suggestion
To execute this file using Python, you can use a code block like this:
```

.. code-block:: python3

   from strawberryfields import StarshipEngine
   from strawberryfields.io import load

   prog = load("chip0.xbb")
   result = eng.run(prog)

>    print(result.samples)

Reviewer 2: I love how concise this is!

Executing your xbb file from the command line
=============================================

> To execute this file from the command line, use the ``starship`` command as follows:

Reviewer: Does the `starship` command get put in the path when you `pip install strawberryfields`?

Author: Yes it does -see setup.py in #146

.. code-block:: console

>    (strawberryfields)$ starship --input test.xbb --output out

Reviewer: suggest `out.txt` so people know exactly what kind of file it is
